### PR TITLE
[#6] [KMM] [Backend] Set up core network layer

### LINKED ISSUE #6: [KMM] [Backend] Set up core network layer
# Why 
We need a core layer to intercept all API calls and responses to use the HTTP services. This layer will be based on [Ktor](https://ktor.io/docs/getting-started-ktor-client.html) - a multiplatform asynchronous HTTP client.

# Acceptance Criteria
- Install Ktor and its dependencies.
- Create a networking layer that will be used in the repository.

### PR DESCRIPTION
**Pull request by phongvhd93**
- Close #6

## What happened 👀

- Implement `buildSrc`
- Implement `BuildKonfig`
- Create `ApiClient` and necessary classes

## Insight 📝

Prepare a core network layer in `shared` module for future backend tasks. Due to the need to use secrets `CLIENT_ID` and `CLIENT_SECRET`, I add `BuildKonfig` implementation to manage sensitive data in `buildKonfig.properties`.

## Proof Of Work 📹

The application will work as normal. 

Screenshot 2023-04-19 at 09 09 02

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
+        google()
+        mavenCentral()
+    }
     dependencies {
-        classpath("org.jetbrains.kotlinx:kover:0.6.1")
+        classpath(Dependency.KOTLIN_GRADLE_PLUGIN)
+        classpath(Dependency.GRADLE)
+        classpath(Dependency.KOTLIN_SERIALIZATION)
+        classpath(Dependency.BUILD_KONFIG)
+		classpath(Dependency.KOVER)
```
Comment (AVI5HEK) [inline build.gradle.kts]: This one needs reformatting.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package co.nimblehq.avishek.phong.kmmic.data.local.datasource
+
+import co.nimblehq.avishek.phong.kmmic.data.remote.model.TokenApiModel
+import com.russhwolf.settings.ExperimentalSettingsApi
+import com.russhwolf.settings.Settings
+import com.russhwolf.settings.serialization.decodeValueOrNull
+import com.russhwolf.settings.serialization.encodeValue
+import kotlinx.serialization.ExperimentalSerializationApi
+
+private const val ACCESS_TOKEN_KEY = "access_token"
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt]: Could we use `KEY` as a prefix? This is a common pattern we follow for snake case properties and resources.
```suggestion
private const val KEY_ACCESS_TOKEN = "access_token"
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    fun save(token: TokenApiModel)
+    fun getToken(): TokenApiModel?
+}
+
+@OptIn(ExperimentalSerializationApi::class, ExperimentalSettingsApi::class)
+class TokenLocalDataSourceImpl(private val settings: Settings): TokenLocalDataSource {
+
+    override fun save(token: TokenApiModel) {
+        settings.encodeValue(TokenApiModel.serializer(), ACCESS_TOKEN_KEY, token)
+    }
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt]: We can use a function expression here as well.
```suggestion
    override fun save(token: TokenApiModel) =
        settings.encodeValue(TokenApiModel.serializer(), ACCESS_TOKEN_KEY, token)
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+                        sendWithoutRequest { request ->
+                            request.url.host == Url(BuildKonfig.BASE_URL).host
+                        }
+
+                        refreshTokens {
+                            tokenRemoteDataSource
+                                .refreshToken(
+                                    RefreshTokenApiBody(
+                                        refreshToken = oldTokens?.refreshToken ?: ""
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt]: We can use `orEmpty()` to make it more expressive.
```suggestion
                                        refreshToken = oldTokens?.refreshToken.orEmpty()
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                            }
+                        }
+
+                        sendWithoutRequest { request ->
+                            request.url.host == Url(BuildKonfig.BASE_URL).host
+                        }
+
+                        refreshTokens {
+                            tokenRemoteDataSource
+                                .refreshToken(
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt:77]: How about this to reduce an indentation level?
```suggestion
                            tokenRemoteDataSource.refreshToken(
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.encodeToJsonElement
+import kotlinx.serialization.json.jsonObject
+import co.nimblehq.avishek.phong.kmmic.BuildKonfig
+import co.nimblehq.avishek.phong.kmmic.domain.model.AppError
+
+fun HttpRequestBuilder.path(path: String) {
+    url(BuildKonfig.BASE_URL.plus(path))
+}
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt]: Optional: expression body for one-liner looks nicer.
```suggestion
fun HttpRequestBuilder.path(path: String) = url(BuildKonfig.BASE_URL.plus(path))
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+inline fun <reified T> HttpRequestBuilder.setQueryParameters(parameters: T) {
+    try {
+        val queryParameters = Json.encodeToJsonElement(parameters).jsonObject
+        for ((key, value) in queryParameters) {
+            parameter(key, value)
+        }
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt]: Optional: we can simplify it like this:
```suggestion
        queryParameters.forEach {
            parameter(it.key, it.value)
        }
```

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/RefreshTokenApiBody.kt [hunk 8]
```diff
@@ -0,0 +1,19 @@
+package co.nimblehq.avishek.phong.kmmic.data.remote.body
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+import co.nimblehq.avishek.phong.kmmic.BuildKonfig
+
+private const val GRANT_TYPE_REFRESH_TOKEN = "refresh_token"
+
+@Serializable
+data class RefreshTokenApiBody(
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/RefreshTokenApiBody.kt]: Could we rename it to `RefreshTokenRequestBody` or simply `RefreshTokenRequest`? The same principle applies to the response body as well.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+interface TokenRemoteDataSource {
+
+    fun refreshToken(body: RefreshTokenApiBody): Flow<TokenApiModel>
+}
+
+class TokenRemoteDataSourceImpl(private  val apiClient: ApiClient) : TokenRemoteDataSource {
+
+    override fun refreshToken(body: RefreshTokenApiBody): Flow<TokenApiModel> {
+        return apiClient.responseBody(
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: Why not create the `HttpRequestBuilder` inside the `responseBody` extension? We could pass the `path` and `method` to the method and let it build `HttpRequestBuilder`. 🤔

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt:1]: We should rename the file to `HttpRequestBuilderExt.kt`.

File: build.gradle.kts [hunk 10]
```diff
@@ -1,16 +1,27 @@
...
-    kotlin("multiplatform").version("1.8.0").apply(false)
+    id(Plugin.KOVER_FULL_PATH) version (Version.KOVER)
+}
+
+allprojects {
+    repositories {
+        google()
+        mavenCentral()
+    }
 }
```
Comment (AVI5HEK) [inline build.gradle.kts]: How about we add the JSON API dependency using the Gradle script instead of manually adding the module? Would be easier to manage the versions later on IMO. Or is there a quick automated way to add it to the project? 🤔 
```suggestion
allprojects {
    val buildProperties = rootDir.loadGradleProperties("buildKonfig.properties")
    repositories {
        google()
        mavenCentral()
        maven {
            name = "Github Packages"
            url = uri("https://maven.pkg.github.com/nimblehq/jsonapi-kotlin")
            credentials {
                username = buildProperties.getProperty("GITHUB_USER")
                password = buildProperties.getProperty("GITHUB_TOKEN")
            }
        }
    }
}
```

File: buildSrc/src/main/kotlin/appPackage/Configurations.kt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+object Flavors {
+    const val PRODUCTION = "production"
+    const val STAGING = "staging"
+}
+
+object Module {
+    const val JSONAPI_CORE = ":jsonapi-kotlin:core"
```
Comment (AVI5HEK) [inline buildSrc/src/main/kotlin/appPackage/Configurations.kt]: Then update the dependency. 
```suggestion
    const val JSONAPI_CORE = [":jsonapi-kotlin:core"](co.nimblehq.jsonapi:core:${Version.JSON_API})
```
We should move `JSONAPI_CORE` to `Dependency.kt` in this case.

File: settings.gradle.kts [hunk 12]
```diff
@@ -6,13 +6,7 @@ pluginManagement {
...
-    repositories {
-        google()
-        mavenCentral()
-    }
-}
-
 rootProject.name = "KMM-IC"
 include(":androidApp")
 include(":shared")
+include(":jsonapi-kotlin:core")
```
Comment (AVI5HEK) [inline settings.gradle.kts]: Then we can remove this line,

File: shared/build.gradle.kts [hunk 13]
```diff
@@ -28,21 +32,46 @@ kotlin {
...
+                implementation(Dependency.KTOR_AUTH)
+
+                implementation(Dependency.NAPIER)
+
+                // settings
+                implementation(Dependency.SETTINGS)
+                implementation(Dependency.SETTINGS_SERIALIZATION)
+
+                //Json API
+                implementation(project(Module.JSONAPI_CORE))
```
Comment (AVI5HEK) [inline shared/build.gradle.kts]: And then updating this and syncing the project should work.
```suggestion
                implementation(Dependency.JSONAPI_CORE)
```

**Review by AVI5HEK — COMMENTED** [review]
Left a few comments. Looking forward to knowing your thoughts.

**Comment by phongvhd93** [comment]
> Left a few comments. Looking forward to knowing your thoughts.

I'll resolve them and let you know. Because we still need to setup Koin in this PR, so I will change this PR to draft.

File: build.gradle.kts (hunk 1 above)
Comment (phongvhd93) [reply build.gradle.kts]: That is so weird; I see it as normal on my side.

Screenshot 2023-04-21 at 09 56 21

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt (hunk 5 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt:77]: I think breaking the new line at `.refreshToken` will make it look consistent with `.last()` and `run` below.

File: build.gradle.kts (hunk 10 above)
Comment (phongvhd93) [reply build.gradle.kts]: Let me try this, I faced issues with version compatibility with JSON API, so I decided to fork to my repo and solve the issues. We also set our app minSdk = 21 but in the JSON API, it is 23.

Screenshot 2023-04-21 at 10 11 38

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt (hunk 4 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/ApiClient.kt]: solved in commit #a2a4693

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt (hunk 3 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt]: solved in commit a2a4693

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt (hunk 2 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/local/datasource/TokenLocalDataSource.kt]: solved in commit a2a4693

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/HttpRequestBuilder.kt:1]: solved in commit a2a4693

File: .gitmodules [hunk 14]
```diff
@@ -0,0 +1,3 @@
+[submodule "jsonapi-kotlin"]
+	path = jsonapi-kotlin
+	url = git@github.com:phongvhd93/nimble-jsonapi-kotlin.git
```
Comment (suho) [inline .gitmodules]: Let's use https://github.com/nimblehq/jsonapi-kotlin, if there is issue with minimum version, we can update our project instead

Comment (phongvhd93) [reply .gitmodules]: @suho, I solved in ead4a4a

File: androidApp/build.gradle.kts
Comment (kaungkhantsoe) [inline androidApp/build.gradle.kts:1]: Please change `kotlinCompilerExtensionVersion` to `1.3.1` [Ref](https://developer.android.com/jetpack/androidx/releases/compose-kotlin) 🙏 . Not sure how you guys can build the app, I cannot build the project without changing it.

Comment (phongvhd93) [reply androidApp/build.gradle.kts:1]: @kaungkhantsoe, I solved this in c4d5617

File: .github/workflows/ios_automatic_code_review.yml [hunk 15]
```diff
@@ -13,13 +17,22 @@ jobs:
           working-directory: shared
       steps:
       - uses: actions/checkout@v2
+        with:
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/ios_automatic_code_review.yml]: Please remove this

File: .github/workflows/ios_automatic_code_review.yml [hunk 16]
```diff
@@ -45,6 +53,8 @@ jobs:
 
       - name: Checkout
         uses: actions/checkout@v3
+        with:
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/ios_automatic_code_review.yml]: Please remove this

File: .github/workflows/release_ios_appstore.yml [hunk 17]
```diff
@@ -17,13 +17,22 @@ jobs:
           working-directory: shared
       steps:
       - uses: actions/checkout@v2
+        with:
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/release_ios_appstore.yml]: Please remove this

File: .github/workflows/release_ios_appstore.yml [hunk 18]
```diff
@@ -44,6 +53,9 @@ jobs:
 
       - name: Checkout
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/release_ios_appstore.yml]: Please remove this

File: .github/workflows/release_ios_staging.yml [hunk 19]
```diff
@@ -17,13 +17,22 @@ jobs:
           working-directory: shared
       steps:
       - uses: actions/checkout@v2
+        with:
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/release_ios_staging.yml]: Please remove this

File: androidApp/build.gradle.kts [hunk 20]
```diff
@@ -9,7 +9,7 @@ android {
     compileSdk = 33
     defaultConfig {
         applicationId = "co.nimblehq.avishek.phong.kmmic.android"
-        minSdk = 21
+        minSdk = 23
```
Comment (AVI5HEK) [inline androidApp/build.gradle.kts:12]: I actually asked [this](https://nimble-co.slack.com/archives/C051TNKTDFB/p1682052334167079?thread_ts=1682047801.791059&cid=C051TNKTDFB) after seeing this change here. Since JSON API now supports 21, can't we keep it 21 here as well? 🤔

File: .github/workflows/release_ios_staging.yml [hunk 21]
```diff
@@ -44,6 +53,9 @@ jobs:
 
       - name: Checkout
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+          submodules: recursive
```
Comment (suho) [inline .github/workflows/release_ios_staging.yml]: Same here, please remove this

File: .gitignore [hunk 22]
```diff
@@ -125,4 +125,5 @@ buildSrc/src/main/kotlin/appPackage/BuildKonfig.kt
 
 # unless supporting rvm < 1.11.0 or doing something fancy, ignore this:
 .rvmrc
+buildKonfig.properties
```
Comment (suho) [inline .gitignore:128]: It's better to have a file `buildKonfig.properties.example`

**Review by AVI5HEK — COMMENTED** [review]
One more concern regarding the min SDK.

File: settings.gradle.kts [hunk 23]
```diff
@@ -6,13 +6,7 @@ pluginManagement {
...
-    repositories {
-        google()
-        mavenCentral()
-    }
-}
-
 rootProject.name = "KMM-IC"
 include(":androidApp")
 include(":shared")
+//include(":jsonapi-kotlin:core")
```
Comment (suho) [inline settings.gradle.kts]: Please remove this line

File: shared/shared.podspec [hunk 24]
```diff
@@ -36,4 +36,4 @@ Pod::Spec.new do |spec|
         }
     ]
                 
-end
+end
```
Comment (suho) [inline shared/shared.podspec]: Please add new line here, you can config in Android Studio.

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/RefreshTokenApiBody.kt (hunk 8 above)
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/body/RefreshTokenApiBody.kt]: @AVI5HEK @phongvhd93 Based on Google doc (https://developer.android.com/topic/architecture/data-layer#business-models), it's better to use `ApiModel` or `ApiBody`, it will be more clear. `RefreshTokenRequestBody` also work with me.

File: shared/shared.podspec (hunk 24 above)
Comment (kaungkhantsoe) [inline shared/shared.podspec]: Missing eof

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/model/TokenApiModel.kt [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+) {
+    constructor(token: Token) : this(
+        "-",
+        "-",
+        token.accessToken,
+        token.tokenType,
+        token.expiresIn,
+        token.refreshToken,
+        token.createdAt
+    )
```
Comment (kaungkhantsoe) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/model/TokenApiModel.kt]: I believe the intention here is for converting `Token` to `TokenApiModel`.

We usually separate this into another method like `toToken` method you defined below.
```
fun Token.toApiModel() = TokenApiModel(
    id = "",
    type = "",
    accessToken = accessToken,
    tokenType = tokenType,
    expiresIn = expiresIn,
    refreshToken = refreshToken,
    createdAt = createdAt,
)
```

File: androidApp/build.gradle.kts (hunk 20 above)
Comment (phongvhd93) [reply androidApp/build.gradle.kts:12]: @AVI5HEK, I don't have any problem with this. However, the `shared` module minSdk is 23, and the Android app which consumes the `shared` module has minSdk=21; it can be a compatibility issue that is unsurfaced. What do you think if I keep the minSdk = 23 here?

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/model/TokenApiModel.kt (hunk 25 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/model/TokenApiModel.kt]: Thanks for the advice, let me try that idea

Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/model/TokenApiModel.kt]: @kaungkhantsoe, I improved that in 7d6309e

File: shared/shared.podspec (hunk 24 above)
Comment (phongvhd93) [reply shared/shared.podspec]: solved in 7d6309e

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+
+data class Token(
+    val accessToken: String,
+    val tokenType: String,
+    val expiresIn: Int,
+    val refreshToken: String,
+    val createdAt: Int
+)
+
+fun Token.toApiModel() = TokenApiModel(
```
Comment (kaungkhantsoe) [inline shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt]: This `toApiModel` should be inside `TokenApiModel`. According to Clean architecture `domain` layer doesn't have access to `data` layer. By declaring this here, we are violating the architecture design pattern. 🙏

Comment (kaungkhantsoe) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt]: @phongvhd93 `CheckLoggedInUseCaseImpl` inside `CheckLoggedInUseCase` is also not following the Clean architecture. Please move `CheckLoggedInUseCaseImpl` to `data` layer instead. 🙏

File: androidApp/build.gradle.kts (hunk 20 above)
Comment (AVI5HEK) [reply androidApp/build.gradle.kts:12]: @phongvhd93 Are you talking about this one? It's set to `21` in the `:shared` module. 🤔 
image

File: settings.gradle.kts (hunk 23 above)
Comment (AVI5HEK) [inline settings.gradle.kts]: We can remove this line now.

File: androidApp/build.gradle.kts (hunk 20 above)
Comment (phongvhd93) [reply androidApp/build.gradle.kts:12]: @AVI5HEK, oh sorry, I forgot that we updated the JSON-API to use the one from Nimble's not from the fork from my repo. Let me check it again

Comment (phongvhd93) [reply androidApp/build.gradle.kts:12]: MinSDK must be 23 to use the JSON API. I also updated the `build.gradle.kts` in this PR

Screenshot 2023-04-24 at 17 15 32

File: settings.gradle.kts (hunk 23 above)
Comment (phongvhd93) [reply settings.gradle.kts]: @AVI5HEK solved in cc86dfc

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt (hunk 26 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt]: @kaungkhantsoe solved in cc86dfc

**Review by kaungkhantsoe — APPROVED** [review]
LGTM 🚀 . FYI you can re-request for review after you have updated the PR in the following session. Then the reviewers will get notified again automatically. 😉  
Screenshot 2023-04-24 at 5 23 57 PM

File: androidApp/build.gradle.kts (hunk 20 above)
Comment (AVI5HEK) [reply androidApp/build.gradle.kts:12]: Oh, I see. Let's go with 23 then. ☹️

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt (hunk 26 above)
Comment (AVI5HEK) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt]: @kaungkhantsoe `CheckLoggedInUseCaseImpl` contains the business logic and _is_ part of the domain layer. In fact, we don't need an interface for this, we can mock the concrete implementation.

File: .github/workflows/ios_automatic_code_review.yml [hunk 27]
```diff
@@ -13,13 +17,22 @@ jobs:
           working-directory: shared
       steps:
       - uses: actions/checkout@v2
+        with:
```
Comment (markgravity) [inline .github/workflows/ios_automatic_code_review.yml]: We can revert this change now 🙏 
Please do the same to other workflows 🙌

File: gradle.properties [hunk 28]
```diff
@@ -11,3 +11,6 @@ android.nonTransitiveRClass=true
 #MPP
 kotlin.mpp.enableCInteropCommonization=true
 kotlin.mpp.androidSourceSetLayoutVersion=2
+
+# Default flavor
```
Comment (markgravity) [inline gradle.properties]: Following the other comment convention above, we don't need to add a space here 🙏 
`# Default flavor` -> `#Default flavor`

**Comment by phongvhd93** [comment]
@suho @markgravity I solved your comment in 23e5bb4

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt (hunk 9 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/data/remote/datasource/TokenRemoteDataSource.kt]: solved in 23e5bb4

File: .gitignore (hunk 22 above)
Comment (phongvhd93) [reply .gitignore:128]: solved in 23e5bb4

File: buildKonfig.properties.example.properties [hunk 29]
```diff
@@ -0,0 +1,13 @@
+# Staging env
+STAGING_CLIENT_ID=R9v5TscbY-_gxCtcmeJHIPm5tzXKkopGS**********
```
Comment (markgravity) [inline buildKonfig.properties.example.properties]: Are they real data? 🙈

**Review by markgravity — APPROVED** [review]
Please check other comments 🙏

File: buildSrc/src/main/kotlin/appPackage/Configurations.kt (hunk 11 above)
Comment (phongvhd93) [reply buildSrc/src/main/kotlin/appPackage/Configurations.kt]: I updated it in 7fcdc128

Screenshot 2023-04-25 at 15 47 56

File: shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt (hunk 26 above)
Comment (phongvhd93) [reply shared/src/commonMain/kotlin/co/nimblehq/avishek/phong/kmmic/domain/model/Token.kt]: > CheckLoggedInUseCaseImpl contains the business logic and is part of the domain layer
- So the use case should be in domain layer following [this](https://developer.android.com/topic/architecture/domain-layer). I will update it
> In fact, we don't need an interface for this; we can mock the concrete implementation.
- Using an interface will make our code loose coupling with orther part.
- Secondly, I think using interface will help us easier to mock with Mockative and less manual work.

File: buildKonfig.properties.example.properties (hunk 29 above)
Comment (phongvhd93) [reply buildKonfig.properties.example.properties]: Half of real is not real :D You can't use it or guess it. Just shed light on how to use this file.

**Review by AVI5HEK — APPROVED** [review]
LGTM. 🎉
